core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -3200,8 +3209,41 @@ private boolean o3Commit(long lag) {
             // Safe check of the sort. No known way to reproduce
             assert o3TimestampMin <= o3TimestampMax;
 
-            if (lag > 0) {
-                long lagThresholdTimestamp = o3TimestampMax - lag;
+            if (o3MaxLag > 0) {
+                long lagError = 0;
+                if (getMaxTimestamp() != Long.MIN_VALUE) {
+
+                    // When table already has data we can calculate the overlap of the newly added
+                    // batch of records with existing data in the table. Positive value of the overlap
+                    // means that our o3MaxLag was undersized.
+
+                    lagError = getMaxTimestamp() - o3CommitBatchTimestampMin;
+
+                    int n = WINDOW_SIZE - 1;
+
+                    if (lagError > 0) {
+                        o3EffectiveLag += lagError * 1.5;
+                        o3EffectiveLag = Math.min(o3EffectiveLag, o3MaxLag);
+                    } else {
+                        // avoid using negative effective o3MaxLag
+                        o3EffectiveLag += Math.max(0, lagError * 0.5);
```

Comment:
`lagError` is <= 0 here, in this else branch. This line always equevalent to `o3EffectiveLag += 0;`

Reply:
fixed here: https://github.com/questdb/questdb/pull/2759/commits/91471b8e2f6e20f0aa15345ca6f77c75b9f08ce8